doc/design/backfill-starvation-prevention.md

```diff
@@ -0,0 +1,179 @@
+ 
+## Algorithm
+
+Kube batch scheduler does scheduling in rounds. In each round, it runs the following four scheduling actions in order: reclaim, allocate, preempt, backfill. The actual action sequence is configurable. 
+
+This section describes the gist of the algorithm. To visualize how the algorithm works, please take a look at [the slides](https://docs.google.com/presentation/d/1kWbR0_lRGSjkgCnUoySNUktIqKVcR2OOTsyq7WtUIGI/edit#slide=id.g509241a485_0_1041).
+
+### Allocation Phase
+* In allocation phase, we allocate resources on a node to a task, if either of the following two conditions is true
+  * The job is starvating. If a job is starving, we will allocate the remaining resources (of a node) to the task even if the node’s remaining resource is less than the task’s request resource. By doing so, we prevent the scheduler from allocating the remaining resources to other jobs. In other words, we reserve the remaining resources for the starving job. Eventually, some running jobs on the node will finish running, and the starving task will get enough resource to start. This is how we prevent a job from be starved for ever.
+  * The task’s request resource is less than the node’s accessible resource. Remember that accessible resource include idle resource and backfilled resource. If a task’s request resource is less than a node’s accessible resource, it means that the task should have enough resource to start if backfill tasks finish running or being evicted. 
+* At the end allocation phase, we do the following things based on a job’s state. 
+  * We call dispatcher to bind pod to node if the job is in ready state. In other words, we will start the job. 
+  * We do nothing if the job is not in ready state.
+
+### Backfill Phase
+* If a job is starving, we do nothing. We will not release resources allocated to starving tasks. In other words, we will not backfill tasks into resources owned by a starving tasks. The resources be allocated for the starving task until the task gets enough resources to start.
+
+* If a job is ready or conditionally ready, we do nothing. We will not release resources allocated to ready or conditionally ready tasks. In other words, we will not backfill tasks into resources owned by ready or conditionally ready tasks. 
+
+* If a job is neither ready nor conditionally ready, we will release resources allocated to it. Then we will schedule, or backfill, other tasks into the released resources.
+
+
+### Notes
+
+The reservation based starvation prevention design described above has one assumption - it assumes that the order of nodes is stable across scheduling rounds. If the scheduler sees nodes in one order, say [n1, n2, …, nm], in round x, it expects to see the nodes in the same order in round x+1. Starvation may happen if node order is not stable. The implementation will ensure the order is stable.
+
+## References
+
+### Kube-batch Architecture
+![kube-batch-arch](../images/kube-batch-arch.png)
```

Comment:
`kube-batch-arch.png` to `Kube-batch-arch.png` or rename the picture.

Reply:
done.